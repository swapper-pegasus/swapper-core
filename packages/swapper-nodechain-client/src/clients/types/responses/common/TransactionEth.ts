export type TransactionEth = {
    blockHash: string,
    blockNumber: string,
    from: string,
    gas: string,
    gasPrice: string,
    hash: string,
    input: string,
    nonce: string,
    r: string,
    s: string,
    to: string,
    transactionIndex: string,
    v: string,
    value: string,
}
