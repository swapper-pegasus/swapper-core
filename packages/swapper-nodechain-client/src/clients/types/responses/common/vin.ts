export type vin = {
    coinbase: string,
    sequence: number,
    txinwitness: Array<string>
}
