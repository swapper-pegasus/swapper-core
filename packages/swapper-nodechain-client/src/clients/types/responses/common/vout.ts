type scriptPubKey = {
    addresses: Array<string>,
    asm: string,
    hex: string,
    reqSigs: number,
    type: string,
}

export type vout = {
    n: number,
    scriptPubKey: scriptPubKey,
    value: number,
}
