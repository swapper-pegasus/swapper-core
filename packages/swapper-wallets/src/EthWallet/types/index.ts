export * from './TxObject'
export * from './WalletBalance'
