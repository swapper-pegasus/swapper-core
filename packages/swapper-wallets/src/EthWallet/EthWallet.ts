import {
  SwapperNodechainEthClient,
  ResponseBroadcastTransaction,
  ResponseGetAddressesBalance
} from '@swapper-org/swapper-nodechain-client'
import { Transaction } from '@ethereumjs/tx'
import Common, { Chain } from '@ethereumjs/common'
import { Address, PrefixedHexString, privateToAddress } from 'ethereumjs-util'
import { mnemonicToSeed } from 'bip39'
import hdkey from 'hdkey'
import BN from 'bignumber.js'
import { IWallet } from '../interfaces/IWallet'
import { TxObjectDescription } from './types/TxObject'
import { WalletBalance } from './types/WalletBalance'

export class EthWallet implements IWallet {
  private nodechainClientInstance: SwapperNodechainEthClient
  private ethereumConfig: Common
  private mnemonic: string
  private derivationPath: string

  constructor (
    nodechainServerUrl: string,
    chain: Chain,
    mnemonic: string,
    derivationPath = "m/44'/60'/0'/0/0"
  ) {
    this.nodechainClientInstance = new SwapperNodechainEthClient(
      nodechainServerUrl
    )
    this.ethereumConfig = new Common({ chain })
    this.derivationPath = derivationPath
    this.mnemonic = mnemonic
  }

  private async node () {
    const seed = await mnemonicToSeed(this.mnemonic)
    return hdkey.fromMasterSeed(seed)
  }

  private async hdKey () {
    const node = await this.node()
    return node.derive(this.derivationPath)
  }

  private bnToHex (value: BN): PrefixedHexString {
    return '0x' + value.toString(16)
  }

  public async exportPrivateKey () {
    const { privateKey } = await this.hdKey()
    return privateKey.toString('hex')
  }

  public async transfer (
    txObjectDescription: TxObjectDescription
  ): Promise<ResponseBroadcastTransaction> {
    const hdKey = await this.hdKey()
    const valueHexFormat: PrefixedHexString = this.bnToHex(
      txObjectDescription.value
    )
    const getTransactionCountResponse =
              await this.nodechainClientInstance.getTransactionCount({
                address: txObjectDescription.from,
                pending: true
              })
    const gasPriceResponse =
              await this.nodechainClientInstance.getGasPrice()
    const gasEstimationResponse =
              await this.nodechainClientInstance.estimateGas({
                tx: {
                  gasPrice: gasPriceResponse.result.gasPrice,
                  nonce: getTransactionCountResponse.result.transactionCount,
                  from: txObjectDescription.from,
                  to: txObjectDescription.to,
                  data: txObjectDescription.data || '',
                  value: valueHexFormat
                }
              })

    const finalTransaction = {
      ...txObjectDescription,
      gasPrice: gasPriceResponse.result.gasPrice,
      value: valueHexFormat,
      gasLimit: gasEstimationResponse.result.estimatedGas,
      nonce: Number(getTransactionCountResponse.result.transactionCount)
    }

    const tx = Transaction.fromTxData(finalTransaction, {
      common: this.ethereumConfig
    })

    const signedTx = tx.sign(hdKey.privateKey)
    const rawTransaction: string = signedTx.serialize().toString('hex')

    const result = await this.nodechainClientInstance.broadcastTransaction({
      rawTransaction
    })
    return result
  }

  public async getAddresses (): Promise<string[]> {
    const hdKey = await this.hdKey()
    const address = new Address(privateToAddress(hdKey.privateKey))
    return [address.toString()]
  }

  public async getBalance (): Promise<WalletBalance> {
    const addresses: string[] = await this.getAddresses()
    const balances: ResponseGetAddressesBalance = await this.nodechainClientInstance.getAddressesBalance({ addresses })
    const totalConfirmedBalance: BN = balances.result.reduce((currentBalance, nextBalance) => {
      const balanceConfirmed = new BN(nextBalance.balance.confirmed.replace('0x', ''), 16)
      return currentBalance.plus(balanceConfirmed)
    }, new BN('0'))
    const totalUnconfirmedBalance: BN = balances.result.reduce((currentBalance, nextBalance) => {
      const balanceConfirmed = new BN(nextBalance.balance.unconfirmed.replace('0x', ''), 16)
      return currentBalance.plus(balanceConfirmed)
    }, new BN('0'))
    return { confirmed: totalConfirmedBalance.toString(10), unconfirmed: totalUnconfirmedBalance.toString(10) } // Returns in wei
  }
}
