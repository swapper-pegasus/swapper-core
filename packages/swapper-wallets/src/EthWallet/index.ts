import { EthWallet } from './EthWallet'
import { Chain } from '@ethereumjs/common'
import { WalletBalance, TxObjectDescription } from './types'
export { EthWallet, Chain as EthChain, WalletBalance, TxObjectDescription }
