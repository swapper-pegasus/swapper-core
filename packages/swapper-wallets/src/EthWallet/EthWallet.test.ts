import { EthWallet } from '../EthWallet'
import { Chain } from '@ethereumjs/common'
import BN from 'bn.js'

const nodechainUrlServer = 'http://localhost:7001'
const ethAddress = '0x625ACaEdeF812d2842eFd2Fb0294682A868455bd'

const nmenomic = 'piece effort bind that embrace enrich remind powder sudden patient legend group'
const derivativePath = "m/44'/60'/0'/0/0"

const ethAddressB = '0x93261B4021dbd6200Df9B36B151f4ECF34889e94'

describe('EthWallet', () => {
  it('transfer', async () => {
    const ethWallet = new EthWallet(nodechainUrlServer, Chain.Ropsten, nmenomic, derivativePath)
    const result = await ethWallet.transfer({ from: ethAddress, to: ethAddressB, value: new BN('1000000000000000000'), data: '' })
    expect(result).toMatchSnapshot()
  })
  it('getAddresses', async () => {
    const ethWallet = new EthWallet(nodechainUrlServer, Chain.Ropsten, nmenomic, derivativePath)
    const result = await ethWallet.getAddresses()
    expect(result).toMatchSnapshot()
  })
  it('getBalance', async () => {
    const ethWallet = new EthWallet(nodechainUrlServer, Chain.Ropsten, nmenomic, derivativePath)
    const result = await ethWallet.getBalance()
    expect(result).toMatchSnapshot()
  })
  it('exportPrivateKey', async () => {
    const ethWallet = new EthWallet(nodechainUrlServer, Chain.Ropsten, nmenomic, derivativePath)
    const result = await ethWallet.exportPrivateKey()
    expect(result).toMatchSnapshot()
  })
})
