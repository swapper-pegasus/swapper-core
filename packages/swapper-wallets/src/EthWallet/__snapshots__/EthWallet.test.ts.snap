// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthWallet exportPrivateKey 1`] = `"6fa76995e9a39e852f893e8347c662453a5d517846d150bdf3ddf7601c4bc74c"`;

exports[`EthWallet getAddresses 1`] = `
Array [
  "0x625acaedef812d2842efd2fb0294682a868455bd",
]
`;

exports[`EthWallet getBalance 1`] = `
Object {
  "confirmed": "1",
  "unconfirmed": "0",
}
`;

exports[`EthWallet transfer 1`] = `
Object {
  "id": 1,
  "jsonrpc": "broadcastTransaction",
  "result": Object {
    "broadcasted": "0x213123123123123",
  },
}
`;
